skiko/src/jvmMain/cpp/linux/SoftwareRedrawer.cc

```diff
@@ -0,0 +1,121 @@
+#include <jawt_md.h>
+#include <X11/Xresource.h>
+#include <stdint.h>
+#include "jni_helpers.h"
+
+#include "SkSurface.h"
+#include "src/core/SkAutoMalloc.h"
+
+class SoftwareDevice
+{
+public:
+    Display* display;
+    Window window;
+    GC gc;
+    SkAutoMalloc surfaceMemory;
+    sk_sp<SkSurface> surface;
+    unsigned int depth = 0;
+
+    ~SoftwareDevice() {
+        XFreeGC(display, gc);
+    }
+};
+
+extern "C"
+{
+    void getDisplayDepthBits(SoftwareDevice *device) {
+        Window window;
+        int x, y;
+        unsigned int w, h, border, displayDepth;
+        XGetGeometry(
+            device->display, device->window, &window, &x, &y, &w, &h, &border, &device->depth);
+    }
+
+    SkColorType getDeviceColorSpace(SoftwareDevice *device) {
+        if (device->depth >16) {
+            return kBGRA_8888_SkColorType;
+        }
+        return kRGB_565_SkColorType;
```

Comment:
I'd write it as
switch (device->depth) {
   case 24: case 32: return  kBGRA_8888_SkColorType;
   case 16: return kRGB_565_SkColorType;
   default: assert(false); return XXX;
```